File: frontend/src/api/cursoIntro.ts
import { client } from "@/api/client";

export type CursoIntroCohorteDTO = {
	id: number;
	nombre?: string | null;
	anio_academico: number;
	profesorado_id?: number | null;
	profesorado_nombre?: string | null;
	turno_id?: number | null;
	turno_nombre?: string | null;
	ventana_id?: number | null;
	ventana_tipo?: string | null;
	fecha_inicio?: string | null;
	fecha_fin?: string | null;
	cupo?: number | null;
	observaciones?: string | null;
};

export type CursoIntroCohortePayload = {
	nombre?: string | null;
	anio_academico: number;
	profesorado_id?: number | null;
	turno_id?: number | null;
	ventana_id?: number | null;
	fecha_inicio?: string | null;
	fecha_fin?: string | null;
	cupo?: number | null;
	observaciones?: string | null;
};

export type CursoIntroRegistroDTO = {
	id: number;
	estudiante_id: number;
	estudiante_nombre: string;
	estudiante_apellido: string;
	estudiante_dni: string;
	profesorado_id?: number | null;
	profesorado_nombre?: string | null;
	cohorte_id?: number | null;
	cohorte_nombre?: string | null;
	turno_id?: number | null;
	turno_nombre?: string | null;
	resultado: string;
	resultado_display: string;
	asistencias_totales?: number | null;
	nota_final?: number | null;
	observaciones?: string | null;
	es_historico: boolean;
	resultado_at?: string | null;
};

export type CursoIntroRegistroFilters = {
	cohorte_id?: number;
	profesorado_id?: number;
	turno_id?: number;
	resultado?: string;
	anio?: number;
};

export type CursoIntroPendienteDTO = {
	estudiante_id: number;
	estudiante_dni: string;
	estudiante_nombre: string;
	estudiante_apellido: string;
	profesorados: Array<{
		id: number | null;
		nombre: string;
		anio_ingreso?: number | null;
	}>;
	anio_ingreso?: number | null;
};

type CursoIntroVentanaDTO = {
	id: number;
	desde: string;
	hasta: string;
	activo: boolean;
	periodo?: string | null;
};

export type CursoIntroEstadoDTO = {
	aprobado: boolean;
	registro_actual?: CursoIntroRegistroDTO | null;
	cohortes_disponibles: CursoIntroCohorteDTO[];
	ventanas: CursoIntroVentanaDTO[];
};

export async function fetchCursoIntroCohortes(): Promise<
	CursoIntroCohorteDTO[]
> {
	const { data } = await client.get<CursoIntroCohorteDTO[]>(
		"/estudiantes/curso-intro/cohortes",
	);
	return data;
}

export async function crearCursoIntroCohorte(
	payload: CursoIntroCohortePayload,
): Promise<CursoIntroCohorteDTO> {
	const { data } = await client.post<CursoIntroCohorteDTO>(
		"/estudiantes/curso-intro/cohortes",
		payload,
	);
	return data;
}

export async function actualizarCursoIntroCohorte(
	cohorteId: number,
	payload: CursoIntroCohortePayload,
): Promise<CursoIntroCohorteDTO> {
	const { data } = await client.put<CursoIntroCohorteDTO>(
		`/estudiantes/curso-intro/cohortes/${cohorteId}`,
		payload,
	);
	return data;
}

export async function listarCursoIntroRegistros(
	filters: CursoIntroRegistroFilters = {},
): Promise<CursoIntroRegistroDTO[]> {
	const params: Record<string, number | string> = {};
	if (filters.cohorte_id) params.cohorte_id = filters.cohorte_id;
	if (filters.profesorado_id) params.profesorado_id = filters.profesorado_id;
	if (filters.turno_id) params.turno_id = filters.turno_id;
	if (filters.resultado) params.resultado = filters.resultado;
	if (filters.anio) params.anio = filters.anio;
	const { data } = await client.get<CursoIntroRegistroDTO[]>(
		"/estudiantes/curso-intro/registros",
		{ params },
	);
	return data;
}

export async function listarCursoIntroPendientes(
	profesoradoId?: number,
	solo_activos?: boolean,
	anio_ingreso?: number,
	solo_confirmados?: boolean,
): Promise<CursoIntroPendienteDTO[]> {
	const params: Record<string, number | boolean> = {};
	if (profesoradoId) params.profesorado_id = profesoradoId;
	if (solo_activos !== undefined) params.solo_activos = solo_activos;
	if (solo_confirmados !== undefined) params.solo_confirmados = solo_confirmados;
	if (anio_ingreso) params.anio_ingreso = anio_ingreso;
	const { data } = await client.get<CursoIntroPendienteDTO[]>(
		"/estudiantes/curso-intro/pendientes",
		{ params },
	);
	return data;
}

export async function inscribirCursoIntro(payload: {
	cohorte_id: number;
	estudiante_id: number;
	profesorado_id?: number;
	turno_id?: number;
}): Promise<CursoIntroRegistroDTO> {
	const { data } = await client.post<CursoIntroRegistroDTO>(
		"/estudiantes/curso-intro/registros",
		payload,
	);
	return data;
}

export async function registrarCursoIntroAsistencia(
	registroId: number,
	asistencias_totales: number,
): Promise<CursoIntroRegistroDTO> {
	const { data } = await client.post<CursoIntroRegistroDTO>(
		`/estudiantes/curso-intro/registros/${registroId}/asistencia`,
		{ asistencias_totales },
	);
	return data;
}

export async function cerrarCursoIntroRegistro(
	registroId: number,
	payload: {
		nota_final?: number;
		asistencias_totales?: number;
		resultado: string;
		observaciones?: string;
	},
): Promise<CursoIntroRegistroDTO> {
	const body: Record<string, unknown> = { resultado: payload.resultado };
	if (payload.nota_final !== undefined) body.nota_final = payload.nota_final;
	if (payload.asistencias_totales !== undefined)
		body.asistencias_totales = payload.asistencias_totales;
	if (payload.observaciones !== undefined)
		body.observaciones = payload.observaciones;
	const { data } = await client.post<CursoIntroRegistroDTO>(
		`/estudiantes/curso-intro/registros/${registroId}/cierre`,
		body,
	);
	return data;
}

export async function fetchCursoIntroEstado(): Promise<CursoIntroEstadoDTO> {
	const { data } = await client.get<CursoIntroEstadoDTO>(
		"/estudiantes/curso-intro/estado",
	);
	return data;
}

export async function autoInscribirCursoIntro(payload: {
	cohorte_id: number;
	profesorado_id?: number | null;
	turno_id?: number | null;
}): Promise<CursoIntroRegistroDTO> {
	const { data } = await client.post<CursoIntroRegistroDTO>(
		"/estudiantes/curso-intro/auto-inscripcion",
		payload,
	);
	return data;
}

File: backend/apps/estudiantes/schemas/curso_intro.py
from datetime import date

from ninja import Field, Schema

# ==========================================
# 8. CURSO INTRODUCTORIO
# ==========================================


class CursoIntroCohorteIn(Schema):
    nombre: str | None = None
    anio_academico: int
    profesorado_id: int | None = None
    turno_id: int | None = None
    ventana_id: int | None = None
    fecha_inicio: date | None = None
    fecha_fin: date | None = None
    cupo: int | None = None
    observaciones: str | None = None


class CursoIntroCohorteOut(Schema):
    id: int
    nombre: str | None = None
    anio_academico: int
    profesorado_id: int | None = None
    profesorado_nombre: str | None = None
    turno_id: int | None = None
    turno_nombre: str | None = None
    ventana_id: int | None = None
    ventana_tipo: str | None = None
    fecha_inicio: str | None = None
    fecha_fin: str | None = None
    cupo: int | None = None
    observaciones: str | None = None


class CursoIntroRegistroOut(Schema):
    id: int
    estudiante_id: int
    estudiante_nombre: str
    estudiante_apellido: str
    estudiante_dni: str
    profesorado_id: int | None = None
    profesorado_nombre: str | None = None
    cohorte_id: int | None = None
    cohorte_nombre: str | None = None
    turno_id: int | None = None
    turno_nombre: str | None = None
    resultado: str
    resultado_display: str
    asistencias_totales: int | None = None
    nota_final: float | None = None
    observaciones: str | None = None
    es_historico: bool
    resultado_at: str | None = None


class CursoIntroRegistroIn(Schema):
    cohorte_id: int | None = None
    estudiante_id: int
    profesorado_id: int | None = None
    turno_id: int | None = None


class CursoIntroAsistenciaIn(Schema):
    asistencias_totales: int = Field(ge=0, le=100)


class CursoIntroCierreIn(Schema):
    nota_final: float | None = Field(default=None, ge=1, le=10)
    asistencias_totales: int | None = Field(default=None, ge=0, le=100)
    resultado: str
    observaciones: str | None = None


class CursoIntroPendienteOut(Schema):
    estudiante_id: int
    estudiante_dni: str
    estudiante_nombre: str
    estudiante_apellido: str
    profesorados: list[dict] = Field(default_factory=list)
    anio_ingreso: int | None = None


class CursoIntroVentanaOut(Schema):
    id: int
    desde: str
    hasta: str
    activo: bool
    periodo: str | None = None


class CursoIntroEstadoOut(Schema):
    aprobado: bool
    registro_actual: CursoIntroRegistroOut | None = None
    cohortes_disponibles: list[CursoIntroCohorteOut] = Field(default_factory=list)
    ventanas: list[CursoIntroVentanaOut] = Field(default_factory=list)


CursoIntroEstadoOut.model_rebuild()


class CursoIntroAutoInscripcionIn(Schema):
    cohorte_id: int
    profesorado_id: int | None = None
    turno_id: int | None = None
